compiler/ir/backend.jvm/entrypoint/src/org/jetbrains/kotlin/backend/jvm/JvmIrFragmentCodegenFactory.kt

```diff
@@ -0,0 +1,225 @@
+import org.jetbrains.kotlin.resolve.jvm.JvmClassName
+import org.jetbrains.kotlin.resolve.source.PsiSourceFile
+import org.jetbrains.kotlin.serialization.deserialization.descriptors.DeserializedContainerSource
+
+// Copied and modified from JvmIrCodegenFactory. The vision is that, as we
+// understand the requirements of the fragment codegen more I feel we may be
+// able to simplify this/accommodate the fragment codegen better.
+class JvmIrFragmentCodegenFactory(
+    private val configuration: CompilerConfiguration,
+    private val codegenInfo: CodeFragmentCodegenInfo,
+    private val phaseConfig: PhaseConfig,
+) : CodegenFactory {
+
+    private data class JvmIrBackendInput(
+        val state: GenerationState,
+        val irModuleFragment: IrModuleFragment,
+        val symbolTable: SymbolTable,
+        val phaseConfig: PhaseConfig,
+        val irProviders: List<IrProvider>,
+        val extensions: JvmGeneratorExtensionsImpl,
+        val backendExtension: JvmBackendExtension,
+        val notifyCodegenStart: () -> Unit,
+    )
+
+    override fun generateModule(state: GenerationState, files: Collection<KtFile>) {
+        val input = convertToIr(state, files)
+        doGenerateFilesInternal(input)
+    }
+
+    private fun translateCodegenInfo(codegenInfo: CodeFragmentCodegenInfo): EvaluatorFragmentInfo {
+        return EvaluatorFragmentInfo(codegenInfo.classDescriptor, codegenInfo.methodDescriptor, codegenInfo.parameters.map { it.targetDescriptor })
+    }
+
+    private fun convertToIr(state: GenerationState, files: Collection<KtFile>): JvmIrBackendInput {
+        val mangler = JvmDescriptorMangler(MainFunctionDetector(state.bindingContext, state.languageVersionSettings))
+        val symbolTable =
+            FragmentCompilerSymbolTableDecorator(JvmIdSignatureDescriptor(mangler), IrFactoryImpl, JvmNameProvider, translateCodegenInfo(codegenInfo))
+        val psi2ir = Psi2IrTranslator(state.languageVersionSettings, Psi2IrConfiguration())
+        val messageLogger = state.configuration[IrMessageLogger.IR_MESSAGE_LOGGER] ?: IrMessageLogger.None
+        val jvmGeneratorExtensions = object : JvmGeneratorExtensionsImpl(configuration) {
+            override fun getContainerSource(descriptor: DeclarationDescriptor): DeserializedContainerSource? {
+                val containingFile =
+                    (descriptor.toSourceElement.containingFile as? PsiSourceFile)?.psiFile as? KtFile ?: return super.getContainerSource(
+                        descriptor
+                    )
+                val fileClassInfo = getFileClassInfoNoResolve(containingFile)
+                val fileClassName = JvmClassName.byFqNameWithoutInnerClasses(fileClassInfo.fileClassFqName)
+                val facadeClassName = JvmClassName.byFqNameWithoutInnerClasses(fileClassInfo.facadeClassFqName)
+                return JvmPackagePartSource(fileClassName, facadeClassName, ProtoBuf.Package.getDefaultInstance(), object : NameResolver {
```

Comment:
Maybe create another `DeserializedContainerSource` implementation instead?

If there's some code in the backend which expects it to be exactly `JvmPackagePartSource`, I suppose we should introduce an intermediate interface with the needed parts between `JvmPackagePartSource` and `DeserializedContainerSource`, use it in that backend code, and implement it here

Reply:
Did a take on this, let me know if I got your concerns right.

I'll admit I don't understand the full importance or mechanism of `DeserializedContainerSource`, but I introduced an interface for the backend (in `DeclarationStubGenerator`) to depend on, rather than explicitly on `JvmPackagePartSource`.

Then, I introduced a new implementation of `DeserializedContainerSource` letting me move some of this code into there. 

I added a bit of explanation for this to the code as well.

Reply:
Yes, this is exactly what I proposed, thanks!